internal/impl/kube.go

```diff
@@ -514,7 +536,24 @@ scrape_configs:
 
 // buildDeployment generates a kubernetes deployment for a replica set.
 func buildDeployment(rs *replicaSetInfo, dep *protos.Deployment, image string) (*v1.Deployment, error) {
-	name := name{dep.App.Name, rs.name, dep.Id[:8]}.DNSLabel()
+	// Unique name that persists across app versions.
+	globalName := name{dep.App.Name, rs.name}.DNSLabel()
+
+	// Per deployment name that is app version specific.
+	depName := name{dep.App.Name, rs.name, dep.Id[:8]}.DNSLabel()
+
+	var oname string
+	matchLabels := map[string]string{}
+	podLabels := map[string]string{"depName": depName, "metrics": dep.App.Name}
+	if hasListenersFn(rs) {
+		oname = globalName
+		matchLabels["globalName"] = globalName
```

Comment:
I'm a bit confused about how an internal Service will talk to this Deployment, since we don't set the "depName" for it (and the internal Service matches based on "depName").

Could you try to run an example where "main" and some other component Foo are co-located, and a third component talks to the Foo component? I'm not sure if we have any such example.

Reply:
We need to add labels for deployment and global to the pod labels. We already do that, so i think it works? I added a TODO to test with main though, because we don't have an app that does that.